### DISCUSSION TITLE
Requests don't go to the S3 server and processed inside the container

### DISCUSSION THREAD

**Discussion by dmrlx** [nginxinc/nginx-s3-gateway#186] [Q&A] [answered]
I'm running nginx-s3-gateway with the following command:

```docker run --env-file ./settings.env -p80:80 ghcr.io/nginxinc/nginx-s3-gateway/nginx-oss-s3-gateway:latest```

settings.env file content:

```
S3_BUCKET_NAME=test-bucket
AWS_ACCESS_KEY_ID=testuser
AWS_SECRET_ACCESS_KEY=111111111
S3_SERVER=10.23.23.32
S3_SERVER_PORT=9000
S3_SERVER_PROTO=https
S3_REGION=us-east-1
S3_STYLE=path
AWS_SIGS_VERSION=4
ALLOW_DIRECTORY_LIST=false
```

Then I open the following link:

```http://10.23.23.33/path/to/file.txt```

and get 404 error.

In log I see following lines:

```
[root@host ~]# docker run --env-file ./settings.env --publish 80:80 ghcr.io/nginxinc/nginx-s3-gateway/nginx-oss-s3-gateway:latest-njs-oss
/docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration
/docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
/docker-entrypoint.sh: Ignoring /docker-entrypoint.d/00-check-for-required-env.sh, not executable
/docker-entrypoint.sh: Ignoring /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh, not executable
/docker-entrypoint.sh: Ignoring /docker-entrypoint.d/15-local-resolvers.envsh
/docker-entrypoint.sh: Ignoring /docker-entrypoint.d/20-envsubst-on-templates.sh, not executable
/docker-entrypoint.sh: Ignoring /docker-entrypoint.d/22-enable_js_fetch_trusted_certificate.sh, not executable
/docker-entrypoint.sh: Ignoring /docker-entrypoint.d/30-tune-worker-processes.sh, not executable
/docker-entrypoint.sh: Configuration complete; ready for start up
2023/10/26 14:38:32 [notice] 1#1: using the "epoll" event method
2023/10/26 14:38:32 [notice] 1#1: nginx/1.25.3
2023/10/26 14:38:32 [notice] 1#1: built by gcc 12.2.0 (Debian 12.2.0-14)
2023/10/26 14:38:32 [notice] 1#1: OS: Linux 3.10.0-1160.102.1.el7.x86_64
2023/10/26 14:38:32 [notice] 1#1: getrlimit(RLIMIT_NOFILE): 1048576:1048576
2023/10/26 14:38:32 [notice] 1#1: start worker processes
2023/10/26 14:38:32 [notice] 1#1: start worker process 17
2023/10/26 14:38:35 [error] 17#17: *2 open() "/usr/share/nginx/html/path/to/file.txt" failed (2: No such file or directory), client: 10.23.23.31, server: localhost, request: "GET /path/to/file.txt HTTP/1.1", host: "10.23.23.32"
10.23.23.31 - - [26/Oct/2023:14:38:35 +0000] "GET /path/to/file.txt HTTP/1.1" 404 153 "-" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.0 Safari/605.1.15" "-"
```

It looks to me like my request doesn't go to the S3 server and is processed inside the container. We are using minio as S3 backend.

Does anyone have any ideas what I'm doing wrong?

**Comment by alessfg** [comment]
Hey @dmrlx! Everything seems to be working on our end, but it seems as if you are running into an issue with the Docker image startup scripts.  We need a little bit more info before we can try to help more:
* What distro are you using?
* What Docker version are you using?
* Is your OS or Docker configured in such a way that might potentially block executable files from being run?

**Reply by dmrlx** [reply]
Hi, thanks for quick response. Regarding your questions:

- What distro are you using? — CentOS Linux release 7.9.2009 (Core)
- What Docker version are you using? – Docker version 1.13.1, build 7d71120/1.13.1
- Is your OS or Docker configured in such a way that might potentially block executable files from being run? – I have absolutely clean installation of CentOS (with disabled SELinux and firewalld) and absolutely clean installation of docker. Don't sure if anything can block something here.

**Reply by dmrlx** [reply]
Just in case, docker was installed using this instruction: https://phoenixnap.com/kb/how-to-install-docker-centos-7

**Reply by 4141done** [reply]
I don't have an immediate idea of what could be wrong here, but these lines indicate a problem:

```
/docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/
/docker-entrypoint.sh: Ignoring /docker-entrypoint.d/00-check-for-required-env.sh, not executable
/docker-entrypoint.sh: Ignoring /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh, not executable
/docker-entrypoint.sh: Ignoring /docker-entrypoint.d/15-local-resolvers.envsh
/docker-entrypoint.sh: Ignoring /docker-entrypoint.d/20-envsubst-on-templates.sh, not executable
/docker-entrypoint.sh: Ignoring /docker-entrypoint.d/22-enable_js_fetch_trusted_certificate.sh, not executable
/docker-entrypoint.sh: Ignoring /docker-entrypoint.d/30-tune-worker-processes.sh, not executable
```

As a reference, here is what I see starting up with the same image that works:
```
Region: us-east-2

Addressing Style: path

AWS Signatures Version: v4

DNS Resolvers:  192.168.65.5

Directory Listing Enabled: false

Directory Listing Path Prefix: 

Provide Index Pages Enabled: 

Append slash for directory enabled: 

Stripping the following headers from responses: x-amz-;

CORS Enabled: 0

/docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh

10-listen-on-ipv6-by-default.sh: info: Getting the checksum of /etc/nginx/conf.d/default.conf

10-listen-on-ipv6-by-default.sh: info: Enabled listen on IPv6 in /etc/nginx/conf.d/default.conf

/docker-entrypoint.sh: Ignoring /docker-entrypoint.d/15-local-resolvers.envsh

/docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh

20-envsubst-on-templates.sh: Running envsubst on /etc/nginx/templates/default.conf.template to /etc/nginx/conf.d/default.conf

20-envsubst-on-templates.sh: Running envsubst on /etc/nginx/templates/cache.conf.template to /etc/nginx/conf.d/cache.conf

20-envsubst-on-templates.sh: Running envsubst on /etc/nginx/templates/gateway/v4_js_vars.conf.template to /etc/nginx/conf.d/gateway/v4_js_vars.conf

20-envsubst-on-templates.sh: Running envsubst on /etc/nginx/templates/gateway/js_fetch_trusted_certificate.conf.template to /etc/nginx/conf.d/gateway/js_fetch_trusted_certificate.conf

20-envsubst-on-templates.sh: Running envsubst on /etc/nginx/templates/gateway/cors.conf.template to /etc/nginx/conf.d/gateway/cors.conf

20-envsubst-on-templates.sh: Running envsubst on /etc/nginx/templates/gateway/v2_headers.conf.template to /etc/nginx/conf.d/gateway/v2_headers.conf

20-envsubst-on-templates.sh: Running envsubst on /etc/nginx/templates/gateway/v2_js_vars.conf.template to /etc/nginx/conf.d/gateway/v2_js_vars.conf

20-envsubst-on-templates.sh: Running envsubst on /etc/nginx/templates/gateway/v4_headers.conf.template to /etc/nginx/conf.d/gateway/v4_headers.conf

20-envsubst-on-templates.sh: Running envsubst on /etc/nginx/templates/gateway/s3_server.conf.template to /etc/nginx/conf.d/gateway/s3_server.conf

20-envsubst-on-templates.sh: Running envsubst on /etc/nginx/templates/gateway/s3listing_location.conf.template to /etc/nginx/conf.d/gateway/s3listing_location.conf

20-envsubst-on-templates.sh: Running envsubst on /etc/nginx/templates/gateway/s3_location.conf.template to /etc/nginx/conf.d/gateway/s3_location.conf

20-envsubst-on-templates.sh: Running envsubst on /etc/nginx/templates/upstreams.conf.template to /etc/nginx/conf.d/upstreams.conf

/docker-entrypoint.sh: Launching /docker-entrypoint.d/22-enable_js_fetch_trusted_certificate.sh

/docker-entrypoint.sh: Launching /docker-entrypoint.d/30-tune-worker-processes.sh

/docker-entrypoint.sh: Configuration complete; ready for start up
```

I can look in to it a bit more but don't have a similar environment on which I can easily test.

In the mean time, we did just release a new version of the image that uses a newer version of NGINX. As a gut check, you might want to try testing with the old version to see if we have introduced some kind of bug: 
https://github.com/nginxinc/nginx-s3-gateway/pkgs/container/nginx-s3-gateway%2Fnginx-oss-s3-gateway/137972990?tag=latest-njs-oss-20231016

**Reply by dmrlx** [reply]
@4141done Thanks for the reply and the idea. I'll give it a try. Anyway, at least I realise that the "not executable" lines are a symptom of the problem and I can try to investigate further.

**Reply by dmrlx** [reply]
@4141done And one more question. What linux distribution and version of docker are you using? I could try installing the exact same environment.

**Reply by 4141done** [reply]
So I'm mostly testing on MacOS which I realize doesn't help. Our [CI builds](https://github.com/nginxinc/nginx-s3-gateway/blob/632bf950db93649c206781944a305b68e642805c/.github/workflows/main.yml#L14) run on `ubuntu:latest` though and that's our normal non MacOS test environment. 

I'm not aware of any quirk of docker that would cause the host OS to matter in regards to the result of a `-x` check but I'll do some digging since it's confusing me.  I appreciate you bringing up the issue and I hope we can work towards a solution together.

**Reply by thresheek** [reply, marked as answer]
This looks similar to https://github.com/nginxinc/docker-nginx/issues/804 and a few others.

Try updating your docker install from 
https://docs.docker.com/engine/install/centos/ please.

**Reply by dmrlx** [reply]
Yeah, it looks like docker upgrade using https://docs.docker.com/engine/install/centos/ instruction helped. Thanks a lot!
